### PR TITLE
Add batch read support, update QUIC to RFC9000 (fixes #5)

### LINKED ISSUE #5: Add batch reading support
Blocks https://github.com/syncthing/syncthing/issues/7824

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
Fixes #5

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
@imsodin could you please review

**Comment by AudriusButkevicius** [comment]
@calmh could you please review

File: conn.go [hunk 1]
```diff
@@ -76,26 +79,109 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
...
+		msgs = append(msgs, msg)
+		if msg.Err != nil {
+			return 0, msg.Err
+		}
+	case <-r.closed:
+		return 0, errClosed
+	}
+
+	// After that, it's best effort. If there are messages, we read them.
+	// If not, we break out and return what we got.
```
Comment (AudriusButkevicius) [inline conn.go:141]: In theory there are `syscall.MSG_WAITONE/WAITALL` flags, but that's unix only, and I am not bothered to add support for these.

The usecase here is quic, and that doesn't use any flags.

File: conn.go [hunk 2]
```diff
@@ -76,26 +79,109 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
...
+		}
+
+		ms[i].N = msg.N
+		ms[i].NN = msg.NN
+		ms[i].Flags = msg.Flags
+		ms[i].Addr = msg.Addr
+
+		copy(ms[i].Buffers[0], msg.Buffers[0][:msg.N])
+
+		// Truncate OOB data.
```
Comment (AudriusButkevicius) [inline conn.go:172]: Is this right? No idea, but we do this in the other case too.

File: filter.go [hunk 3]
```diff
@@ -25,15 +34,28 @@ type Config struct {
...
 // connection.
 func NewPacketFilter(conn net.PacketConn) *PacketFilter {
+	// This is derived from quic codebase.
+	var batchSize = 0
+	switch runtime.GOOS {
+	case "linux":
+		batchSize = 8
+	case "freebsd", "darwin":
+		batchSize = 1
+	}
```
Comment (AudriusButkevicius) [inline filter.go:53]: In theory this is not needed, seems go does the right thing anyway.

**Review by calmh — APPROVED** [review]
Seems correct enough to my understanding. A couple of things that strike me that might be potential improvements or not:

- The underlying reader loop uses a buffer pool and stuff, and `ReadBatch` copies each message for the caller. It seems like this could be simplified to removing the pooling and just allocating directly in the read loop instead and returning those buffers directly, avoiding some work and some copying...

- The message copying code is repeated with magic handling for (buggy?) length fields around OOB data and whatnot, which seems like it could be a `copyMessage(*socket.Message) *socket.Message` or something once and for all (unless we can avoid the copying...)

**Comment by AudriusButkevicius** [comment]
> The underlying reader loop uses a buffer pool and stuff, and ReadBatch copies each message for the caller. It seems like this could be simplified to removing the pooling and just allocating directly in the read loop instead and returning those buffers directly, avoiding some work and some copying...

Why would that be avoiding work and allocations?

The real read loop is independent of the callers reading from the sockets, hence we need to allocate to fill data that is coming off the socket.

Once the fake socket user calls a read method, then we need to copy the data into the buffers of the caller, at which point the buffers we got from the pool during underlying read can go back. 

Sure, we could not use a pool and let the GC collect them, but that's not less work, and less copying, it's just more strain for the GC.

**Comment by AudriusButkevicius** [comment]
Simplified the second point (I think?)

**Comment by AudriusButkevicius** [comment]
@calmh bump re comments, perhaps I am missing something.

File: conn.go [hunk 4]
```diff
@@ -76,24 +79,113 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
...
+
+	var timeout <-chan time.Time
+
+	if deadline, ok := r.deadline.Load().(time.Time); ok && !deadline.IsZero() {
+		timer := time.NewTimer(deadline.Sub(time.Now()))
+		timeout = timer.C
+		defer timer.Stop()
+	}
+
+	var msgs []messageWithError
```
Comment (imsodin) [inline conn.go:110]:
```suggestion
	msgs := make([]messageWithError, 0, len(ms))
```

File: conn.go [hunk 5]
```diff
@@ -76,24 +79,113 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
...
+			return 0, errNotSupported
+		}
+
+		n, nn, err := copyBuffers(msg, ms[i].Buffers[0], ms[i].OOB)
+		if err != nil {
+			return 0, err
+		}
+
+		ms[i].N = n
+		ms[i].N = nn
```
Comment (imsodin) [inline conn.go:160]:
```suggestion
		ms[i].NN = nn
```

File: conn.go [hunk 6]
```diff
@@ -76,24 +79,113 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
...
+	copy(buf, msg.Buffers[0][:msg.N])
+
+	// Truncate, probably?
+	oobn := msg.NN
+	if oobl := len(oobBuf); oobl < oobn {
+		oobn = oobl
+	}
+	if oobn > 0 {
+		copy(oobBuf, msg.OOB[:oobn])
+	}
```
Comment (imsodin) [inline conn.go:186]: Why no error here if the provided oob buffer is too small?

File: filter.go (hunk 3 above)
Comment (imsodin) [reply filter.go:53]: Should we then better rely on what go does so we stay up to date when changes happen?

File: filter.go [hunk 7]
```diff
@@ -139,28 +169,70 @@ func (d *PacketFilter) Overflow() uint64 {
...
+	result := make([]messageWithError, n)
+
+	for i := 0; i < n; i++ {
+		result[i].Err = err
+		result[i].Message = batch[i]
+	}
+
+	for _, msg := range batch[n:] {
+		d.returnBuffers(msg)
+	}
```
Comment (imsodin) [inline filter.go:228]: Aren't buffers returned in outer functions (`loop`, `ReadFrom`, ...)?

File: conn.go (hunk 6 above)
Comment (AudriusButkevicius) [reply conn.go:186]: This is just a random decision, with no good merit or particular reason.

File: filter.go (hunk 3 above)
Comment (AudriusButkevicius) [reply filter.go:53]: This avoids using ReadBatch on Windows, which is not supported anyway.

File: filter.go (hunk 7 above)
Comment (AudriusButkevicius) [reply filter.go:228]: Yes, the ones that we return as part of result (`for i := 0; i < n; i++ {`), everything after n has to be returned by us, as we are not returning that back to the caller.

**Comment by AudriusButkevicius** [comment]
Not sure why, but the reply to `Should we then better rely on what go does so we stay up to date when changes happen?` is not visible in non-file view.

**Review by imsodin — APPROVED** [review]

**Pull request merged**